File: app/src/main/java/remix/myplayer/ui/activity/LibraryActivity.java
package remix.myplayer.ui.activity;

import android.os.Bundle;
import java.util.List;
import remix.myplayer.ui.misc.MultipleChoice;
import remix.myplayer.ui.adapter.BaseAdapter;
import remix.myplayer.util.Constants;

/**
 * Created by Remix on 2017/10/20.
 */

public abstract class LibraryActivity<D, A extends BaseAdapter> extends MenuActivity implements
    android.app.LoaderManager.LoaderCallbacks<List<D>> {

  protected A mAdapter;
  protected MultipleChoice<D> mChoice = new MultipleChoice<>(this, Constants.SONG);

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    if (mHasPermission) {
      getLoaderManager().initLoader(getLoaderId(), null, this);
    }
  }

  protected abstract int getLoaderId();


  @Override
  public void onMediaStoreChanged() {
    super.onMediaStoreChanged();
    if (mHasPermission) {
      getLoaderManager().restartLoader(getLoaderId(), null, this);
    } else {
      if (mAdapter != null) {
        mAdapter.setData(null);
      }
    }
  }

  @Override
  public void onPermissionChanged(boolean has) {
    if (has != mHasPermission) {
      mHasPermission = has;
      onMediaStoreChanged();
    }
  }

  @Override
  public void onPlayListChanged(String name) {

  }

  @Override
  public android.content.Loader<List<D>> onCreateLoader(int id, Bundle args) {
    return getLoader();
  }

  @Override
  public void onLoadFinished(android.content.Loader<List<D>> loader, List<D> data) {
    if (mAdapter != null) {
      mAdapter.setData(data);
    }
  }

  @Override
  public void onLoaderReset(android.content.Loader<List<D>> loader) {
    if (mAdapter != null) {
      mAdapter.setData(null);
    }
  }

  protected android.content.Loader<List<D>> getLoader() {
    return null;
  }

  @Override
  public void onBackPressed() {
    if (mChoice.isActive()) {
      mChoice.close();
    } else {
      finish();
    }
  }
}

File: app/src/main/java/remix/myplayer/ui/activity/LibraryActivity.kt
package remix.myplayer.ui.activity

import android.app.LoaderManager
import android.content.Loader
import android.os.Bundle
import remix.myplayer.ui.adapter.BaseAdapter
import remix.myplayer.ui.misc.MultipleChoice
import remix.myplayer.util.Constants

/**
 * Created by Remix on 2017/10/20.
 */
abstract class LibraryActivity<Data, Adapter : BaseAdapter<Data, *>?> : MenuActivity(), LoaderManager.LoaderCallbacks<List<Data>> {
  @JvmField
  protected var mAdapter: Adapter? = null
  protected var mChoice = MultipleChoice<Data>(this, Constants.SONG)

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    if (hasPermission) {
      loaderManager.initLoader(loaderId, null, this)
    }
  }

  protected abstract val loaderId: Int
  protected abstract val loader: Loader<List<Data>>

  override fun onMediaStoreChanged() {
    super.onMediaStoreChanged()
    if (hasPermission) {
      loaderManager.restartLoader(loaderId, null, this)
    } else {
      mAdapter?.setDataList(null)
    }
  }

  override fun onPermissionChanged(has: Boolean) {
    if (has != hasPermission) {
      hasPermission = has
      onMediaStoreChanged()
    }
  }

  override fun onPlayListChanged(name: String) {}

  override fun onLoadFinished(loader: Loader<List<Data>>, data: List<Data>?) {
    mAdapter?.setDataList(data)
  }

  override fun onLoaderReset(loader: Loader<List<Data>>) {
    mAdapter?.setDataList(null)
  }

  override fun onCreateLoader(id: Int, args: Bundle?): Loader<List<Data>> {
    return loader
  }

  override fun onBackPressed() {
    if (mChoice.isActive) {
      mChoice.close()
    } else {
      finish()
    }
  }
}